### PR TITLE
Paged lists

### LINKED ISSUE #31: Support paging in list views
We reuse the ListPopup for the changelog popup because it contains all features needed to show a list of text. But scrolling requires moving the selected line to the bottom before any scrolling occurs.

changelog

These lists should be configurable to _not_ show the selected row, and to page the view up and down rather than move line by line.

### PR DESCRIPTION
**Pull request by natecraddock**
This adds paging to list views. Paged list views are purely for showing text content, so there is no selected row, and pressing Enter closes the layer. Scrolling up and down moves all lines.

Also includes a cleanup on the layers code for drawing list lines.

Closes #31

### CODE REVIEW

File: zygrader/ui/components.py [hunk 1]
```diff
@@ -669,6 +671,9 @@ def set_searchable(self, prompt: str, search_fn: Callable):
     def set_sortable(self):
         self._sortable = True
 
+    def set_paged(self):
+        self._paged = True
```
Comment (teiknsmith) [inline zygrader/ui/components.py:675]: It feels weird to make this irreversible, even if there aren't any current use cases for enabling paging and subsequently undoing it.
```suggestion
    def set_paged(self, paged=True):
        self._paged = paged
```

Comment (teiknsmith) [reply zygrader/ui/components.py:675]: Looking everywhere, we would want to do this in Layers as well, with the default param there, so we could make this a requried param here.
```suggestion
    def set_paged(self, paged):
        self._paged = paged
```

File: zygrader/ui/layers.py [hunk 2]
```diff
@@ -579,7 +580,15 @@ def set_searchable(self, prompt: str, search_fn=__string_search_fn):
     def set_sortable(self):
         self.component.set_sortable()
 
+    def set_paged(self):
+        self.component.set_paged()
+        self._paged = True
```
Comment (teiknsmith) [inline zygrader/ui/layers.py:585]: As mentioned in components, reversibility could make sense.
```suggestion
    def set_paged(self, paged=True):
        self.component.set_paged(paged)
        self._paged = paged
```

File: zygrader/main.py [hunk 3]
```diff
@@ -108,7 +108,8 @@ def view_changelog():
     lines = config.versioning.load_changelog()
 
     popup = ui.layers.ListLayer("Changelog", popup=True)
-    popup.set_exit_text("Close")
+    popup.set_exit_text("Press enter to close")
```
Comment (teiknsmith) [inline zygrader/main.py:111]: Technically, escape works to close too. I really don't think we need to say that here, but if you can think of a concise way to include it, we could.
I do like this better than "Press enter or escape to close" just because adding "escape" makes it way too wordy.

File: zygrader/ui/components.py [hunk 4]
```diff
@@ -696,6 +701,18 @@ def _sort(self, lines):
         lines.sort(key=lambda line: line.sort_index)
         return lines
 
+    def _draw_list_lines(self, window, lines, y_start, x_start):
+        for line_number, line in enumerate(lines):
+            prefix = ScrollableList.UNSELECTED_PREFIX
+            attributes = 0
+            if line_number + self._scroll == self._selected_index and not self._paged:
+                prefix = ScrollableList.SELECTED_PREFIX
+                attributes = curses.A_BOLD
```
Comment (teiknsmith) [inline zygrader/ui/components.py]: Just because that conditional is getting long, a function abstraction will help to keep it readable.
```suggestion
        def should_show_selected(line_number):
            is_selected = line_number + self._scroll == self._selected_index
            return is_selected and not self._paged

        for line_number, line in enumerate(lines):
            prefix = ScrollableList.UNSELECTED_PREFIX
            attributes = 0
            if should_show_selected(line_number):
                prefix = ScrollableList.SELECTED_PREFIX
                attributes = curses.A_BOLD
```

File: zygrader/main.py (hunk 3 above)
Comment (teiknsmith) [reply zygrader/main.py:111]: Oh, and also left. But again, it would probably be more verbose than helpful unless we can find a concise way to say "Press enter or escape or left (if you enable left/right arrow navigation) to close"

**Review by teiknsmith — COMMENTED** [review]
For the reversibility of pagedness, I'm really not sure if it would be better with my suggestions or just as is. It would also be easy to add in later if we do find a use case for it.

Also, with the name "paged lists" I intuitively expected the 'PgUp' and 'PgDn' keys to scroll by page. We wouldn't need this, but I think it's worth considering. If we do add events for these keys, they could work in `ListLayers` whether or not they are paged. I'm not a big user of those keys so I'm okay with everything as is, but would welcome the added functionality.

File: zygrader/main.py (hunk 3 above)
Comment (natecraddock) [reply zygrader/main.py:111]: I think only mentioning Enter here is fine. Those who enable `ESC` for going back know it exists, and left arrow becomes natural for those who use it.

File: zygrader/ui/components.py (hunk 1 above)
Comment (natecraddock) [reply zygrader/ui/components.py:675]: Fair enough. I can't imagine any case where we would want to change it though

Comment (teiknsmith) [reply zygrader/ui/components.py:675]: Yeah, I started turning that way by the end of looking at this, and even more so now think that the suggestion above doesn't actually make sense to add.

File: zygrader/ui/layers.py (hunk 2 above)
Comment (teiknsmith) [reply zygrader/ui/layers.py:585]: And as mentioned elsewhere, there's no use case for it, so we don't need to (and perhaps even *shouldn't*) add this suggestion.

**Comment by natecraddock** [comment]
@spencerteiknsmith I think adding page up and page down support is great, but better in a separate commit. I'll commit this now if you don't have any other comments

**Review by teiknsmith — APPROVED** [review]
